source/simulator/solver_schemes.cc

```diff
@@ -958,8 +958,12 @@ namespace aspect
         // Restore particles through stored copy of particle handler,
         // but only if they have already been displaced in a nonlinear
         // iteration (in the assemble_and_solve_composition call).
-        if ((particle_world.get() != nullptr) && (nonlinear_iteration > 0))
-          particle_world->restore_particles();
+
+        for (auto &particle_world : particle_worlds)
+          {
+            if (nonlinear_iteration > 0)
+              particle_world->restore_particles();
+          }
```

Comment:
Just a style question, but I would write:
```suggestion
    if (nonlinear_iteration > 0)
        for (auto &particle_world : particle_worlds)
              particle_world->restore_particles();
```